hydra-node/src/Hydra/Network/Etcd.hs

```diff
@@ -0,0 +1,513 @@
+  KnownHydraVersions (..),
+  Network (..),
+  NetworkCallback (..),
+  NetworkComponent,
+  NetworkConfiguration (..),
+  PortNumber,
+  hydraVersionedProtocolNumber,
+ )
+import Network.GRPC.Client (
+  Address (..),
+  ConnParams (..),
+  Connection,
+  ReconnectPolicy (..),
+  ReconnectTo (ReconnectToOriginal),
+  Server (..),
+  rpc,
+  withConnection,
+ )
+import Network.GRPC.Client.StreamType.IO (biDiStreaming, nonStreaming)
+import Network.GRPC.Common (GrpcError (..), GrpcException (..), HTTP2Settings (..), NextElem (..), def, defaultHTTP2Settings)
+import Network.GRPC.Common.NextElem (whileNext_)
+import Network.GRPC.Common.Protobuf (Protobuf, defMessage, (.~))
+import Network.GRPC.Etcd (KV, Lease, Watch)
+import System.Directory (createDirectoryIfMissing, listDirectory, removeFile)
+import System.FilePath ((</>))
+import System.IO.Error (isDoesNotExistError)
+import System.Posix (Handler (Catch), installHandler, sigTERM)
+import System.Process.Typed (
+  createPipe,
+  getStderr,
+  proc,
+  setStderr,
+  stopProcess,
+  waitExitCode,
+  withProcessTerm,
+ )
+import UnliftIO (readTVarIO)
+
+-- | Concrete network component that broadcasts messages to an etcd cluster and
+-- listens for incoming messages.
+withEtcdNetwork ::
+  (ToCBOR msg, FromCBOR msg, Eq msg) =>
+  Tracer IO EtcdLog ->
+  HydraVersionedProtocolNumber ->
+  -- TODO: check if all of these needed?
+  NetworkConfiguration ->
+  NetworkComponent IO msg msg ()
+withEtcdNetwork tracer protocolVersion config callback action = do
+  -- TODO: fail if cluster config / members do not match --peer
+  -- configuration? That would be similar to the 'acks' persistence
+  -- bailing out on loading.
+  withProcessTerm etcdCmd $ \p -> do
+    -- Ensure the sub-process is also stopped when we get asked to terminate.
+    _ <- installHandler sigTERM (Catch $ stopProcess p) Nothing
+    -- TODO: error handling (also kill threads)
+    race_ (waitExitCode p >>= \ec -> fail $ "etcd exited with: " <> show ec) $ do
+      race_ (traceStderr p) $ do
+        -- TODO: cleanup reconnecting through policy if other threads fail
+        doneVar <- newTVarIO False
+        -- NOTE: The connection to the server is set up asynchronously; the
+        -- first rpc call will block until the connection has been established.
+        withConnection (connParams doneVar) server $ \conn -> do
+          (`finally` atomically (writeTVar doneVar True)) $
+            race_ (pollConnectivity conn localHost callback) $ do
+              race_ (waitMessages conn protocolVersion persistenceDir callback) $ do
+                queue <- newPersistentQueue (persistenceDir </> "pending-broadcast") 100
+                race_ (broadcastMessages tracer conn protocolVersion port queue) $
+                  action
+                    Network
+                      { broadcast = writePersistentQueue queue
+                      }
+ where
+  connParams doneVar =
```

Comment:
Could we have local types on this and others? It's hard to see which module this type comes from.

Reply:
I'd agree if it would not be as straight-forward. Even in low-tech environment like github diff view here shows me the right type and module if I just ctrl+f "connParams".